src/main/java/com/amazon/opendistroforelasticsearch/security/resolver/IndexResolverReplacer.java

```diff
@@ -316,26 +284,61 @@ public boolean replace(final TransportRequest request, boolean retainMode, Strin
         }, false);
     }
 
+    private static final class IndexResolveKey {
+        private final IndicesOptions opts;
+        private final boolean isSearchOrFieldCapabilities;
+        private final String[] original;
+        public IndexResolveKey(IndicesOptions opts, boolean isSearchOrFieldCapabilities, String[] original) {
+            this.opts = opts;
+            this.isSearchOrFieldCapabilities = isSearchOrFieldCapabilities;
+            this.original = original;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            IndexResolveKey that = (IndexResolveKey) o;
+            return isSearchOrFieldCapabilities == that.isSearchOrFieldCapabilities &&
+                    Objects.equal(opts, that.opts) &&
+                    Arrays.equals(original, that.original);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hashCode(opts, isSearchOrFieldCapabilities) + 31*Arrays.hashCode(original);
```

Comment:
nit: not sure if `Objects.hashCode()` is the best option here. I would rather use `opts.hashCode()` and `Boolean.hashCode(isSearchOrFieldCapabilities)`.